modules/compression/src/main/java/org/apache/ignite/spi/communication/LZ4Engine.java

```diff
@@ -0,0 +1,122 @@
+ * limitations under the License.
+ */
+
+package org.apache.ignite.spi.communication;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.lz4.LZ4Exception;
+import net.jpountz.lz4.LZ4Factory;
+import net.jpountz.lz4.LZ4SafeDecompressor;
+import org.apache.ignite.internal.util.nio.compression.CompressionEngine;
+import org.apache.ignite.internal.util.nio.compression.CompressionEngineResult;
+
+import static org.apache.ignite.internal.util.nio.compression.CompressionEngineResult.BUFFER_OVERFLOW;
+import static org.apache.ignite.internal.util.nio.compression.CompressionEngineResult.BUFFER_UNDERFLOW;
+import static org.apache.ignite.internal.util.nio.compression.CompressionEngineResult.OK;
+
+/**
+ * Implementation of LZ4 algorithm.
+ */
+public final class LZ4Engine implements CompressionEngine {
+    /** */
+    private final LZ4Compressor compressor;
+
+    /** */
+    private final LZ4SafeDecompressor decompressor;
+
+    /** */
+    public LZ4Engine() {
+        LZ4Factory factory = LZ4Factory.fastestInstance();
+
+        compressor = factory.fastCompressor();
+        decompressor = factory.safeDecompressor();
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompressionEngineResult compress(ByteBuffer src, ByteBuffer buf) {
+        assert src != null;
+        assert buf != null;
+
+        try {
+            int compress = compressor.compress(src, src.position(), src.remaining(),
+                buf, buf.position() + 4, buf.remaining() - 4);
+
+            putInt(compress, buf);
+
+            buf.position(buf.position() + compress);
+            src.position(src.position() + src.remaining());
+        }
+        catch (LZ4Exception e) {
+            return BUFFER_OVERFLOW;
+        }
+
+        return OK;
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompressionEngineResult decompress(ByteBuffer src, ByteBuffer buf) {
+        assert src != null;
+        assert buf != null;
+
+        int len = src.remaining();
+        int initPos = src.position();
+
+        if (len < 5)
+            return BUFFER_UNDERFLOW;
+
+        int compressedLen = getInt(src);
+
+        assert compressedLen >= 0;
+
+        if (src.remaining() < compressedLen) {
+            src.position(initPos);
+
+            return BUFFER_UNDERFLOW;
+        }
+
+        try {
+            int decompress = decompressor.decompress(src, src.position(), compressedLen,
+                buf, buf.position(), buf.remaining());
+
+            buf.position(buf.position() + decompress);
+            src.position(src.position() + compressedLen);
+        }
+        catch (LZ4Exception e) {
+            src.position(initPos);
+
+            return BUFFER_OVERFLOW;
+        }
+
+        return OK;
+    }
+
+    /** */
+    private static int getInt(ByteBuffer buf) {
+        return ((buf.get() & 0xFF) << 24) | ((buf.get() & 0xFF) << 16)
```

Comment:
Too magic. Add normal javadoc description. And for `putInt` too. Also name of function a bit general.

Reply:
Done.